src/interface.jl

```diff
@@ -39,6 +39,8 @@ end
 # default all rules to first order calls
 apply!(o, state, x, dx, dxs...) = apply!(o, state, x, dx)
 
+isnumeric(x::AbstractFloat) = true
+isnumeric(x::Complex{<:AbstractFloat}) = true
```

Comment:
Maybe it's cleaner to say what's excluded:
```suggestion
isnumeric(x::Number) = true
isnumeric(x::Integer) = false
```
Then when someone wants Unitful numbers, they may just work.

If this is going to be applied everywhere, not as an opt-in, then I think it must exclude integers.

Reply:
Maybe, though if in some distant future Flux is actually capable of quantization-aware training you might well see int8/int4 arrays. RE Unitful quantities, is there any reason to consider them any "safer" to optimize than integers?

Reply:
I think the graphics people do some things with low-bit numbers, but wrap them up to disguise the integer inside --- presumably it would make sense to do something like that before trying to AD with Int8?
```
julia> g = Gray{ColorTypes.N0f8}(128/255)
Gray{N0f8}(0.502)

julia> 0 < g < 1
true

julia> g isa Integer
false

julia> dump(g)
Gray{FixedPointNumbers.N0f8}
  val: FixedPointNumbers.N0f8
    i: UInt8 0x80
```
> RE Unitful quantities, is there any reason to consider them any "safer" to optimize than integers?

My thinking is that if you add `float(zero(x))` to them, which is what some trivial gradient could produce, nothing much will happen. Whereas if you do this to a size or an index, it'll break.

Reply:
Here's a thought: if the concern is about indices and bools, can we just exclude Int64, Int32 and Bool? I've not seen anyone using those types as model params, and it would eliminate 99% of the confusion with indices + true/false.

Reply:
After playing around for a bit with Unitful, I'm still not comfortable with giving carte blanche to all non-integer number types. Even with floats, it's not clear that a scalar param should  be trainable by default.